library/Icingadb/Util/PerfData.php

```diff
@@ -578,15 +611,23 @@ public function formatLabel(bool $html = false): string
 
     public function toArray(): array
     {
+        $min = isset($this->minValue) && !$this->isPercentage()
+            ? $this->format($this->minValue)
+            : '';
+        $max = isset($this->maxValue) && !$this->isPercentage()
+            ? $this->format($this->maxValue)
+            : '';
+        if ($this->isValid === false) {
+            $parts = explode(';', $this->perfdataValue);
+            $min = $min === '' && isset($parts[3]) ? $parts[3] : $min;
+            $max = $max === '' && isset($parts[4]) ? $parts[4] : $max;
+        }
+
         return array(
             'label' => $this->getLabel(),
             'value' => $this->format($this->getvalue()),
-            'min' => isset($this->minValue) && !$this->isPercentage()
-                ? $this->format($this->minValue)
-                : '',
-            'max' => isset($this->maxValue) && !$this->isPercentage()
-                ? $this->format($this->maxValue)
-                : '',
+            'min' => $min,
+            'max' => $max,
```

Comment:
toArray() shouldn't parse. That's the task of parse().

And please reconsider that validate() accepts a parameter to set the Perfdata's isValid flag.

The `raw` properties for example might only be set in case something is invalid, so if any of them is set, the perfdata is invalid.

```suggestion
        return array(
            'label' => $this->getLabel(),
            'value' => isset($this->value) ? $this->format($this->value) : $this->rawValue,
            'min' => (string) ! $this->isPercentage()
                ? (isset($this->minValue) ? $this->format($this->minValue) : $this->rawMinValue)
                : null,
            'max' => (string) ! $this->isPercentage()
                ? (isset($this->maxValue) ? $this->format($this->maxValue) : $this->rawMaxValue)
                : null,
```